torch/nested/_internal/offload_tensor.py

```diff
@@ -0,0 +1,225 @@
+        if self.host_tensor is None:
+            assert self.device_tensor is not None
+            self.host_tensor = self.device_tensor.to("cpu", non_blocking=True)
+            if self.offload_hook is not None:
+                ret = self.offload_hook(self.host_tensor, self.device_tensor)
+                assert ret is None
+        self.device_tensor = None
+        return self.host_tensor
+
+    def restore(self):
+        if self.device_tensor is None:
+            assert self.host_tensor is not None
+            # Do not cache the device_tensor to avoid mutating subclass fields
+            # during compile.
+            ret = self.host_tensor.to(self.device, non_blocking=True)
+            return ret
+        else:
+            return self.device_tensor
+
+    def is_offloaded(self):
+        return self.device_tensor is None
+
+    def __tensor_flatten__(self):
+        ctx = {
+            "device": self.device,
+        }
+        # We specialize on the offloadedness of input tensors to the compiled
+        # region and trace the restore logic into the graph.
+        inner_tensors = []
+        if self.device_tensor is not None:
+            inner_tensors.append("device_tensor")
+        if self.host_tensor is not None:
+            inner_tensors.append("host_tensor")
+        return inner_tensors, ctx
+
+    @staticmethod
+    def __tensor_unflatten__(inner_tensors, meta, outer_size, outer_stride):
+        device_tensor = inner_tensors.get("device_tensor", None)
+        host_tensor = inner_tensors.get("host_tensor", None)
+        device = meta["device"]
+        return create_offload_tensor(
+            device=device, device_tensor=device_tensor, host_tensor=host_tensor
+        )
+
+    @classmethod
+    def __torch_dispatch__(cls, func, types, args, kwargs):
+        # Doing any operation on a CachedTensor automatically calls restore and then unwraps
+        kwargs = {} if kwargs is None else kwargs
+        unwrapped_args = pytree.tree_map_only(
+            OffloadTensor, lambda x: x.restore(), args
+        )
+        unwrapped_kwargs = pytree.tree_map_only(
+            OffloadTensor, lambda x: x.restore(), kwargs
+        )
+        out = func(*unwrapped_args, **unwrapped_kwargs)
+
+        return out
+
+torch.serialization.add_safe_globals([OffloadTensor])
+
+
+_global_is_pending_offload = False
+
+
+def request_offload_all():
+    # Before the next creation of a new offload tensor, we want to offload all
+    # existing offload tensors.
+    # If you request offload all in the compiled region, the offloading will
+    # happen after compiled region is done.
+    global _global_is_pending_offload
+    _global_is_pending_offload = True
+
+
+class OffloadTensorRegistry:
+    def __init__(self, offload_hook: Optional[Callable] = None):
+        self.wrapper_refs = []
+        # Upon execution, the hook
+        # hook(host_tensor, device_tensor) -> None
+        self.offload_hook = None
```

Comment:
why set this to None and not self.offload_hook = offload_hook?

Also, I don't see testing for the offload_hook--how is it meant to be used?